### PR TITLE
Autocomplete username on userinfo page

### PR DESCRIPTION
**Pull request by calledude**
Works with ID as well, although won't autocomplete it per se, it will just show the little dialog when you have entered the full ID for a user.

### CODE REVIEW

File: Modix/ClientApp/src/components/UserLookup/UserSearch.vue [hunk 1]
```diff
@@ -2,10 +2,16 @@
...
                 <div class="control">
-                    <input :class="inputClass" type="text" placeholder="Enter a User ID..." v-model="userId" />
+                    <Autocomplete @select="selectedUser = $event"
+                                  :serviceCall="userServiceCall" placeholder="Enter a username or an ID">
+                        <template slot-scope="{entry}">
+                            <TinyUserView :user="entry" />
+                        </template>
+                    </Autocomplete>
                 </div>
+                <p class="help is-danger"></p>
```
Comment (Scott-Caldwell) [inline Modix/ClientApp/src/components/UserLookup/UserSearch.vue]: Does this actually show up on the page?

**Comment by calledude** [comment]
@Scott-Caldwell no it does not

File: Modix/ClientApp/src/components/UserLookup/UserSearch.vue (hunk 1 above)
Comment (calledude) [reply Modix/ClientApp/src/components/UserLookup/UserSearch.vue]: no clue dude, that bit was basically just copy paste

**Comment by Scott-Caldwell** [comment]
I mean, I pulled your branch and tried using an ID from someone not in the guild and nothing happened

**Comment by calledude** [comment]
Goddamn cache

You're right, some previous stuff I did just ended up looking really weird so I decided not to do it that way

**Comment by calledude** [comment]
Fixed :)

File: Modix.Services/Core/UserService.cs [hunk 2]
```diff
@@ -92,18 +92,19 @@ public class UserService : IUserService
...
 
-            if (user == null)
+            if (user == null && restUser == null)
                 throw new InvalidOperationException($"Discord user {userId} does not exist");
 
             if (user is IGuildUser guildUser)
                 await TrackUserAsync(guildUser);
 
-            return user;
+            return user ?? restUser;
```
Comment (Scott-Caldwell) [inline Modix.Services/Core/UserService.cs]: So we're always pulling the REST user even though we might never do anything with it? :weird:

File: Modix.Services/Core/UserService.cs [hunk 3]
```diff
@@ -97,13 +97,12 @@ public async Task<IUser> GetUserAsync(ulong userId)
...
                     .GetUserAsync(userId);
 
-            if (user == null)
-                throw new InvalidOperationException($"Discord user {userId} does not exist");
-
             if (user is IGuildUser guildUser)
                 await TrackUserAsync(guildUser);
 
-            return user;
+            return user
```
Comment (jmazouri) [inline Modix.Services/Core/UserService.cs]: I would prefer not to change the implementation of GetUserAsync here. Instead, see my comment on `AutocompleteController.cs`.

File: Modix/Controllers/AutocompleteController.cs [hunk 4]
```diff
@@ -46,10 +48,15 @@ public async Task<IActionResult> AutocompleteUsers(string query)
...
                 : UserGuild.Users
-                    .Where(d => d.Username.OrdinalContains(query))
+                    .Where(d => d.Username.OrdinalContains(query) || d.Id.ToString() == query)
                     .Take(10)
-                    .Select(d => ModixUser.FromSocketGuildUser(d));
+                    .Select(ModixUser.FromIUser);
 
+            if (!result.Any() && ulong.TryParse(query, out var userId))
+            {
+                var user = await _userService.GetUserAsync(userId);
```
Comment (jmazouri) [inline Modix/Controllers/AutocompleteController.cs]: This should be `var user = await _userService.GetUserInformationAsync(UserGuild.Id, userId);`

File: Modix.Services/Core/UserService.cs (hunk 3 above)
Comment (calledude) [reply Modix.Services/Core/UserService.cs]: Why?

Comment (jmazouri) [reply Modix.Services/Core/UserService.cs]: Oh, I didn't realize we had refactored out usage of it entirely. In that case, I'd rather remove it later and continue to use `GetUserInformationAsync`, as that's what's used by the `!info` module.

Comment (calledude) [reply Modix.Services/Core/UserService.cs]: Should I remove it now then? Seeing as nothing else uses it, I might as well, right?

Comment (jmazouri) [reply Modix.Services/Core/UserService.cs]: Sure. If it breaks something, we'll find out in CI :D

Comment (calledude) [reply Modix.Services/Core/UserService.cs]: 10/10 I like that strategy

**Pull request merged**